drop typed GEP calls

vtjnash opened the pull request:
Now that we use LLVM 18, and almost have LLVM 19 support, do cleanup to remove LLVM 15/16 type pointer support. LLVM now slightly prefers that we rewrite our complex GEP to use a simple emit_ptrgep call instead, which is also much simpler for julia to emit also.

giordano commented:
I thought the policy was to support the LLVM in the latest stable version of Julia, which at the moment is LLVM 15 in julia v1.10, and LLVM 16 is in the upcoming julia v1.11.

vtjnash commented:
We killed off typed pointers support, so that already killed off much of the reason to support older versions, and it is just a huge maintenance risk for making new changes (which I am working on).

vtjnash committed:
codegen: simplify GEP emission calls

Rewrite our complex GEP emission to use a simple emit_ptrgep call since
it should be equivalent now that opaque pointers are required.

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

gbaraldi commented on src/cgutils.cpp:
```diff
@@ -130,14 +122,8 @@ static Value *stringConstPtr(
-                                       // AddrSpaceCast in case globals are in non-0 AS
-                                       irbuilder.CreateAddrSpaceCast(gv, gv->getValueType()->getPointerTo(0)),
-                                       Args);
-    setName(emission_context, gep, "string_const_ptr");
```
The name got lost here

gbaraldi commented on src/codegen.cpp:
```diff
@@ -6766,21 +6708,13 @@ static Value *get_current_ptls(jl_codectx_t &ctx)
-            ctx.types().T_size,
-            ct,
-            ConstantInt::get(ctx.types().T_size, offsetof(jl_task_t, world_age) / ctx.types().sizeof_ptr),
-            "world_age");
```
We also lost the name here

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55708#issuecomment-2334964034) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b15100f_vs_ac54799/report.md).

vtjnash commented:
@nanosoldier `runbenchmarks("simd", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55708#issuecomment-2335175364) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/9bb31f6_vs_fa1c6b2/report.md).
